// @torbimusic/core/dist/chord/chord.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chord_1 = require("./chord");
var base_chord_1 = require("./models/base-chord");
var superior_chord_1 = require("./models/superior-chord");
var ChordBuilder = /** @class */ (function () {
    function ChordBuilder() {
        this._additions = [];
        this._suspended = [];
        this._superior = new superior_chord_1.SuperiorChord();
        this._bass = undefined;
    }
    ChordBuilder.from = function (chord) {
        return new ChordBuilder()
            .base(chord.base)
            .additions(chord.additions)
            .suspended(chord.suspended)
            .superior(chord.superior)
            .bass(chord.bass);
    };
    ChordBuilder.prototype.base = function (base) {
        this._base = base;
        return this;
    };
    ChordBuilder.prototype.noteBase = function (note) {
        this._base = new base_chord_1.BaseChord(note, this._base.quality);
        return this;
    };
    ChordBuilder.prototype.additions = function (additions) {
        this._additions = additions;
        return this;
    };
    ChordBuilder.prototype.suspended = function (suspended) {
        this._suspended = suspended;
        return this;
    };
    ChordBuilder.prototype.superior = function (superior) {
        this._superior = superior;
        return this;
    };
    ChordBuilder.prototype.bass = function (bass) {
        this._bass = bass;
        return this;
    };
    ChordBuilder.prototype.build = function () {
        return new chord_1.Chord(this._base, this._additions, this._suspended, this._superior, this._bass);
    };
    return ChordBuilder;
}());
exports.ChordBuilder = ChordBuilder;
