// @torbimusic/core/dist/chord/models/base-chord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var note_qualities_1 = require("../../constants/note-qualities");
var constants_1 = require("../../constants");
var BaseChord = /** @class */ (function () {
    function BaseChord(note, quality) {
        this.note = note;
        this.quality = quality;
    }
    BaseChord.clone = function (baseChord) {
        return new BaseChord(baseChord.note, baseChord.quality);
    };
    BaseChord.prototype.toString = function () {
        return this.note.toString() + this.qualityToString();
    };
    BaseChord.prototype.toStringNaturalNote = function () {
        return this.note.toString(constants_1.Notation.SOUTHERN_EUROPEAN) + this.qualityToString();
    };
    BaseChord.prototype.equals = function (chord) {
        return this.note.equals(chord.note) && this.quality === chord.quality;
    };
    BaseChord.prototype.qualityToString = function () {
        var result = "";
        switch (this.quality) {
            case note_qualities_1.ChordQuality.AUGMENTED:
                result = "aug";
                break;
            case note_qualities_1.ChordQuality.DIMINISHED:
                result = "°";
                break;
            case note_qualities_1.ChordQuality.FIFTH:
                result = "5";
                break;
            case note_qualities_1.ChordQuality.MINOR:
                result = "m";
                break;
        }
        return result;
    };
    return BaseChord;
}());
exports.BaseChord = BaseChord;
