// @torbimusic/core/dist/constants/note-qualities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NoteQuality;
(function (NoteQuality) {
    NoteQuality[NoteQuality["MINOR"] = 0] = "MINOR";
    NoteQuality[NoteQuality["MAJOR"] = 1] = "MAJOR";
    NoteQuality[NoteQuality["AUGMENTED"] = 2] = "AUGMENTED";
    NoteQuality[NoteQuality["DIMINISHED"] = 3] = "DIMINISHED";
})(NoteQuality = exports.NoteQuality || (exports.NoteQuality = {}));
var ChordQuality;
(function (ChordQuality) {
    ChordQuality[ChordQuality["MINOR"] = 0] = "MINOR";
    ChordQuality[ChordQuality["MAJOR"] = 1] = "MAJOR";
    ChordQuality[ChordQuality["AUGMENTED"] = 2] = "AUGMENTED";
    ChordQuality[ChordQuality["DIMINISHED"] = 3] = "DIMINISHED";
    ChordQuality[ChordQuality["FIFTH"] = 4] = "FIFTH";
})(ChordQuality = exports.ChordQuality || (exports.ChordQuality = {}));
var PerfectNoteQuality;
(function (PerfectNoteQuality) {
    PerfectNoteQuality[PerfectNoteQuality["AUGMENTED"] = 0] = "AUGMENTED";
    PerfectNoteQuality[PerfectNoteQuality["DIMINISHED"] = 1] = "DIMINISHED";
    PerfectNoteQuality[PerfectNoteQuality["PERFECT"] = 2] = "PERFECT";
    PerfectNoteQuality[PerfectNoteQuality["FIFTH"] = 3] = "FIFTH";
})(PerfectNoteQuality = exports.PerfectNoteQuality || (exports.PerfectNoteQuality = {}));
