// @torbimusic/core/dist/chord/functions/chord-to-string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("../../constants");
function chordToString(chord) {
    return chord.base.toString()
        + chord.superior.simpleSuperiorToString()
        + additionsToString(chord)
        + suspendedToString(chord)
        + chord.superior.alterationsToString()
        + bassToString(chord);
}
exports.chordToString = chordToString;
function chordToStringNaturalNote(chord) {
    return chord.base.toStringNaturalNote()
        + chord.superior.simpleSuperiorToString()
        + additionsToString(chord)
        + suspendedToString(chord)
        + chord.superior.alterationsToString()
        + bassToStringNaturalNote(chord);
}
exports.chordToStringNaturalNote = chordToStringNaturalNote;
function additionsToString(chord) {
    return chord.additions.reduce(function (acc, curr) { return acc + 'add' + curr; }, "");
}
function suspendedToString(chord) {
    return chord.suspended.reduce(function (acc, curr) { return acc + 'sus' + curr; }, "");
}
function bassToString(chord) {
    return !chord.bass.equals(chord.base.note) ? '/' + chord.bass.toString() : '';
}
function bassToStringNaturalNote(chord) {
    return !chord.bass.equals(chord.base.note) ? '/' + chord.bass.toString(constants_1.Notation.SOUTHERN_EUROPEAN) : '';
}
