// @torbimusic/core/dist/interval/interval.quality.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IntervalQuality;
(function (IntervalQuality) {
    IntervalQuality["MINOR"] = "m";
    IntervalQuality["MAJOR"] = "M";
    IntervalQuality["AUGMENTED"] = "aug";
    IntervalQuality["DIMINISHED"] = "dim";
    IntervalQuality["PERFECT"] = "J";
})(IntervalQuality = exports.IntervalQuality || (exports.IntervalQuality = {}));
var IntervalQualities = /** @class */ (function () {
    function IntervalQualities() {
    }
    IntervalQualities.parse = function (grade, qualityString) {
        var quality;
        var perfect = [1, 4, 5, 8, 11];
        switch (qualityString) {
            case '':
                if (perfect.some(function (perfectGrade) { return perfectGrade === grade; })) {
                    quality = IntervalQuality.PERFECT;
                }
                else {
                    quality = IntervalQuality.MAJOR;
                }
                break;
            case 'aug':
                quality = IntervalQuality.AUGMENTED;
                break;
            case 'dim':
                quality = IntervalQuality.DIMINISHED;
                break;
            case 'J':
                quality = IntervalQuality.PERFECT;
                break;
            case 'm':
                quality = IntervalQuality.MINOR;
                break;
            case 'M':
                quality = IntervalQuality.MAJOR;
                break;
        }
        return quality;
    };
    IntervalQualities.toString = function (quality) {
        return IntervalQuality[quality];
    };
    return IntervalQualities;
}());
exports.IntervalQualities = IntervalQualities;
